x/auth/changepubkey/keeper/keeper.go

```diff
@@ -0,0 +1,150 @@
+package keeper
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/tendermint/tendermint/crypto"
+	"github.com/tendermint/tendermint/libs/log"
+
+	"github.com/cosmos/cosmos-sdk/codec"
+	"github.com/cosmos/cosmos-sdk/store/prefix"
+	storetypes "github.com/cosmos/cosmos-sdk/store/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/auth/changepubkey/types"
+	authkeeper "github.com/cosmos/cosmos-sdk/x/auth/keeper"
+)
+
+// Keeper manages history of public keys per account
+type Keeper struct {
+	key sdk.StoreKey
+	cdc codec.BinaryMarshaler
+	ak  authkeeper.AccountKeeper
+}
+
+// NewKeeper returns a new keeper which manages pubkey history per account.
+func NewKeeper(
+	cdc codec.BinaryMarshaler, key sdk.StoreKey, ak authkeeper.AccountKeeper,
+) Keeper {
+
+	return Keeper{key, cdc, ak}
+}
+
+// "Everytime a key for an address is changed, we will store a log of this change in the state of the chain,
+// thus creating a stack of all previous keys for an address and the time intervals for which they were active.
+// This allows dapps and clients to easily query past keys for an account which may be useful for features
+// such as verifying timestamped off-chain signed messages."
+
+// Logger returns a module-specific logger.
+func (pk Keeper) Logger(ctx sdk.Context) log.Logger {
+	return ctx.Logger().With("module", fmt.Sprintf("x/%s", types.ModuleName))
+}
+
+// GetPubKeyHistory Returns the PubKey history of the account at address by time: involves current pubkey
+func (pk Keeper) GetPubKeyHistory(ctx sdk.Context, addr sdk.AccAddress) ([]types.PubKeyHistory, error) {
+	entries := []types.PubKeyHistory{}
+	if pk.ak.GetAccount(ctx, addr) == nil {
+		return entries, fmt.Errorf("account %s not found", addr.String())
+	}
+	iterator := pk.PubKeyHistoryIterator(ctx, addr)
+	defer iterator.Close()
+	for ; iterator.Valid(); iterator.Next() {
+		entry := types.DecodeHistoricalEntry(iterator.Value())
+		entries = append(entries, entry)
+	}
+	currentEntry := pk.GetCurrentPubKeyEntry(ctx, addr)
+	entries = append(entries, currentEntry)
+	return entries, nil
+}
+
+// GetPubKeyHistoricalEntry Returns the PubKey historical entry at a specific time: involves current pubkey
+func (pk Keeper) GetPubKeyHistoricalEntry(ctx sdk.Context, addr sdk.AccAddress, time time.Time) types.PubKeyHistory {
+	iterator := pk.PubKeyHistoryIteratorAfter(ctx, addr, time)
+	defer iterator.Close()
+	for ; iterator.Valid(); iterator.Next() {
+		entry := types.DecodeHistoricalEntry(iterator.Value())
+		if entry.EndTime.After(time) || entry.EndTime.Equal(time) { // TODO: is this inclusive?
```

Comment:
Yeah, let's make it inclusive because we defined it as "EndTime".  If it was "BeginTime" for the next one, then we would make it exclusive.